core/node/shared/src/genesis.rs

```diff
@@ -186,6 +184,44 @@ pub struct GenesisBatchParams {
     pub rollup_last_leaf_index: u64,
 }
 
+// TODO: add metrics to track latency of each stage
+async fn get_storage_logs_for_genesis_block(
+    storage: &mut Connection<'_, Core>,
+) -> Result<Vec<TreeInstruction<StorageKey>>, GenesisError> {
+    let l1_batch_number = L1BatchNumber(0);
+
+    let touched_slots = storage
+        .storage_logs_dal()
+        .get_touched_slots_for_l1_batch(l1_batch_number)
+        .await
+        .context("cannot fetch touched slots")?;
+
+    let hashed_keys_for_writes: Vec<_> = touched_slots.keys().map(StorageKey::hashed_key).collect();
+    let l1_batches_for_initial_writes = storage
+        .storage_logs_dal()
+        .get_l1_batches_and_indices_for_initial_writes(&hashed_keys_for_writes)
+        .await
+        .context("cannot fetch initial writes batch numbers and indices")?;
```

Comment:
I believe loading `l1_batches_for_initial_writes` and the filtering logic later is not necessary. All non-zero touched slots should be written. Come to think of it, we do get `deduplicated_writes` [during genesis batch initialization](https://github.com/matter-labs/zksync-era/blob/9c6ed838ffcfe6bd8fe157c69fbfe8823826849b/core/lib/zksync_core/src/genesis.rs#L435); it would be optimal just to pass these around and convert them to `TreeInstruction`s.